### DISCUSSION TITLE
Estimating trim envelope and maneuvering envelope

### DISCUSSION THREAD

**Discussion by seanmcleod** [JSBSim-Team/jsbsim#802] [Q&A]
I've added a new Python sample to the repo, see [pull request](https://github.com/JSBSim-Team/jsbsim/pull/801) which demonstrates how to generate a trim envelope for an aircraft using JSBSim.

Here is some sample output for the 737 model at 15,000ft.

![image](https://user-images.githubusercontent.com/824856/210178906-017280a2-5203-43c4-8bac-ab68d60d2b71.png)

The idea for generating the trim envelope was based on seeing one in the following presentation.

[Envelope Protection and Recovery Guidance for Upset Conditions](https://ntrs.nasa.gov/api/citations/20170000836/downloads/20170000836.pdf)

![image](https://user-images.githubusercontent.com/824856/210180981-f5dc7034-40a2-40a3-98a7-5f8826817b98.png)

I'm looking for some input on how to go about using JSBSim to calculate the maneuvering envelope as shown in the following example.

![image](https://user-images.githubusercontent.com/824856/210180992-f158d3f1-9f7d-4530-9e74-e6c5d5a4405a.png)

There are a number of papers that are referenced in the presentation, however they all require payment to access. I did find this description regarding the maneuvering envelope.

> In this context, the preferred interpretation of the safe maneuvering envelope considers reachability from the trim envelope. The stable and controllable trim envelope is considered an a-priori safe set. The backwards reachable set is deﬁned as the set of states from where (at least one point in) the trim envelope can be reached. The forwards reachable set is deﬁned as the set of states which can be reached from (at least one point in) the trim envelope. The safe maneuvering ﬂight envelope is then the intersection between the forwards and backwards reachable sets. This interpretation is illustrated in Fig. 8. In addition to the safe envelope, the backwards reachable set is considered as the survivable ﬂight envelope. After an upset due to damage, turbulence, a wake encounter etc., it is possible to bring the aircraft back to a safe trim condition as long as the current ﬂight condition is situated inside the backwards reachable set.

So for example my understanding is that say there is some turbulence, wake etc. which puts the aircraft at say 200KIAS and a flight path angle of +25 degrees (as shown as the red dot in the image below), then it is possible for the aircraft to recover to the trim envelope within 5s given some set of control inputs (elevator, throttle) during those 5s?

![image](https://user-images.githubusercontent.com/824856/210182018-fd7c8adb-8996-411b-863a-32d562112bff.png)

In terms of the red dot, so we know part of the aircraft state at that point, i.e. IAS of 200kt and a flight path angle of +25 deg. However what about the rest of the state, e.g. alpha, angular rates, control inputs etc.? For example do we have to assume and test against a range of states in terms of alpha, control inputs etc. and prove that for each starting state we can get the aircraft back into the trim envelope in 5s?

**Comment by dpculp** [comment]
Good question.  I was wondering if their scenario assumes one "g".  If so then that should take care of the alpha and angular rates.  What I don't get is what is the starting pitch trim and elevator command.  For instance in the 
red dot scenario are we to assume the stab is trimmed for 200 knots and the elevator command is neutral at the beginning of the maneuver?

**Comment by seanmcleod** [comment]
Yep, unfortunately not really any detailed info in the slides, need to get hold of one of the papers I guess.

In terms of the red dot case and the control positions it's not clear, maybe you can end up at the red dot after a major upset from anywhere in the trim envelope, i.e. with the control positions you had at any of the trim points, e.g. trimmed for 250kt -2 deg of gamma.

**Comment by bcoconni** [comment]
That's an awesome piece of work @seanmcleod 👍 
Definitely a great demo of what JSBSim is capable of.
